<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Ryan R. Christ and Louis J. M. Aslett" />

<meta name="date" content="2023-12-11" />

<title>Reproducing kalis Paper LCT Example</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Reproducing kalis Paper LCT Example</h1>
<h4 class="author">Ryan R. Christ and Louis J. M. Aslett</h4>
<h4 class="date">11 December 2023</h4>



<div id="overview-accompanying-data" class="section level1">
<h1>Overview &amp; Accompanying Data</h1>
<p>This vignette reproduces the distance matrix included in the initial
paper introducing our R package kalis. We infer the haplotype structure
at the lactase gene (LCT) in 1000 Genomes Phase 3 data.</p>
<p>Four data files are required. One is the recombination map
<code>lct.map</code> that accompanies this vignette, although another
recombination map of the user’s choosing may be used. Three files
encoding the required 1000 Genomes haplotypes, <code>lct.hap.gz</code> ,
<code>lct.legend.gz</code>, and <code>lct.sample</code>. All four files
can be found in the <code>example</code> folder at Github repository <a href="https://github.com/louisaslett/kalis-bmc">louisaslett/kalis-bmc</a>.
If the user would like to reconstruct these files directly from 1000
Genomes data, we we provide instructions for doing so using
<code>bcftools</code> at the end of this vignette.</p>
</div>
<div id="run-kalis-in-r" class="section level1">
<h1>Run kalis in R</h1>
<p>Start by declaring the number of cores available for kalis to use in
this example (<code>nthreads</code>) and the data directory that should
be used, <code>data_dir</code>. This data directory must contain the
recombination map <code>lct.map</code>, <code>lct.hap.gz</code> ,
<code>lct.legend.gz</code>, and <code>lct.sample</code>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a>nthreads <span class="ot">&lt;-</span> <span class="fu">as.integer</span>(<span class="dv">4</span>)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>data_dir <span class="ot">&lt;-</span> <span class="st">&quot;./&quot;</span></span></code></pre></div>
<p>No further changes should be needed to the following for
replication.</p>
<p>We load our required libraries for this vignette.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">require</span>(<span class="st">&quot;kalis&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">require</span>(<span class="st">&quot;R.utils&quot;</span>)</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="fu">require</span>(<span class="st">&quot;data.table&quot;</span>)</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="fu">require</span>(<span class="st">&quot;fastcluster&quot;</span>)</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="fu">require</span>(<span class="st">&quot;Matrix&quot;</span>) </span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="fu">require</span>(<span class="st">&quot;viridisLite&quot;</span>)</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="fu">require</span>(<span class="st">&quot;kgp&quot;</span>)</span></code></pre></div>
<p>Now we’re ready to run kalis.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Declare LS Model Parameters</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="do">#########################################</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>neg_log10_Ne <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>neg_log10_mu <span class="ot">&lt;-</span> <span class="dv">4</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="co"># Declare Target Locus</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="do">#########################################</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>gene <span class="ot">&lt;-</span> <span class="st">&quot;lct&quot;</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>gene_target_pos <span class="ot">&lt;-</span> <span class="dv">136608646</span> <span class="co"># rs4988235 in hg19 coordinates</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>pos <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="fu">paste0</span>(data_dir, gene, <span class="st">&quot;.legend.gz&quot;</span>))<span class="sc">$</span>position</span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>target_idx <span class="ot">&lt;-</span> <span class="fu">match</span>(<span class="cn">TRUE</span>, pos <span class="sc">&gt;=</span> gene_target_pos)</span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a><span class="co"># run kalis</span></span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a><span class="do">#########################################</span></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a><span class="fu">CacheHaplotypes</span>(<span class="at">haps =</span> <span class="fu">paste0</span>(data_dir, gene, <span class="st">&quot;.hap.gz&quot;</span>))</span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a></span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a>diff_map <span class="ot">&lt;-</span> <span class="fu">diff</span>(<span class="fu">fread</span>(<span class="fu">paste0</span>(data_dir, gene, <span class="st">&quot;.map&quot;</span>))[[<span class="dv">3</span>]])</span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a>pars <span class="ot">&lt;-</span> <span class="fu">Parameters</span>(<span class="at">rho =</span> <span class="fu">CalcRho</span>(diff_map, <span class="at">s =</span> <span class="dv">10</span><span class="sc">^-</span>neg_log10_Ne), <span class="at">mu =</span> <span class="dv">10</span><span class="sc">^-</span>neg_log10_mu)</span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a>fwd <span class="ot">&lt;-</span> <span class="fu">MakeForwardTable</span>(pars)</span>
<span id="cb3-21"><a href="#cb3-21" tabindex="-1"></a>bck <span class="ot">&lt;-</span> <span class="fu">MakeBackwardTable</span>(pars)</span>
<span id="cb3-22"><a href="#cb3-22" tabindex="-1"></a></span>
<span id="cb3-23"><a href="#cb3-23" tabindex="-1"></a><span class="fu">Forward</span>(fwd, pars, target_idx, <span class="at">nthreads =</span> nthreads)</span>
<span id="cb3-24"><a href="#cb3-24" tabindex="-1"></a><span class="fu">Backward</span>(bck, pars, target_idx, <span class="at">nthreads =</span> nthreads)</span>
<span id="cb3-25"><a href="#cb3-25" tabindex="-1"></a></span>
<span id="cb3-26"><a href="#cb3-26" tabindex="-1"></a>M <span class="ot">&lt;-</span> <span class="fu">DistMat</span>(fwd, bck, <span class="at">type =</span> <span class="st">&quot;raw&quot;</span>, <span class="at">nthreads =</span> nthreads)</span></code></pre></div>
<p>Note <code>M</code> may be plotted directly by calling
<code>plot(M)</code>.</p>
</div>
<div id="post-processing-remake-paper-figure" class="section level1">
<h1>Post-processing remake paper figure</h1>
<p>Here we provide post-processing code to reproduce the heatmap in our
paper from the distance matrix <code>M</code>. We start with a few
helper functions.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>plot_mat <span class="ot">&lt;-</span> <span class="cf">function</span>(x, file, <span class="at">raster =</span> <span class="cn">TRUE</span>, <span class="at">rel_scale =</span> <span class="cn">TRUE</span>) {</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>  temp_col_scale <span class="ot">&lt;-</span> <span class="fu">rev</span>(viridisLite<span class="sc">::</span><span class="fu">viridis</span>(<span class="dv">100</span>))</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>  <span class="cf">if</span>(<span class="sc">!</span>rel_scale){</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>    mx <span class="ot">&lt;-</span> <span class="fu">ceiling</span>(<span class="fu">max</span>(x, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>    <span class="cf">if</span>(mx <span class="sc">&gt;</span> <span class="dv">100</span>) { <span class="fu">stop</span>(<span class="st">&quot;the max entry of x cannot exceed 100 for this plot&#39;s color scale&quot;</span>) }</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>    temp_col_scale <span class="ot">&lt;-</span> temp_col_scale[<span class="dv">1</span><span class="sc">:</span>mx]</span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>  }</span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a>  <span class="fu">cairo_pdf</span>(file)</span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a>  <span class="fu">print</span>(lattice<span class="sc">::</span><span class="fu">levelplot</span>(x[, <span class="fu">ncol</span>(x)<span class="sc">:</span><span class="dv">1</span>],</span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a>                           <span class="at">useRaster =</span> raster,</span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a>                           <span class="at">col.regions =</span> grDevices<span class="sc">::</span><span class="fu">colorRampPalette</span>(temp_col_scale)(<span class="dv">100</span>),</span>
<span id="cb4-15"><a href="#cb4-15" tabindex="-1"></a>                           <span class="at">yaxt =</span> <span class="st">&quot;n&quot;</span>, <span class="at">xaxt =</span> <span class="st">&quot;n&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;&quot;</span>, <span class="at">xaxt =</span> <span class="st">&quot;n&quot;</span>))</span>
<span id="cb4-16"><a href="#cb4-16" tabindex="-1"></a>  <span class="fu">dev.off</span>()</span>
<span id="cb4-17"><a href="#cb4-17" tabindex="-1"></a>}</span>
<span id="cb4-18"><a href="#cb4-18" tabindex="-1"></a></span>
<span id="cb4-19"><a href="#cb4-19" tabindex="-1"></a>interp_hapmap <span class="ot">&lt;-</span> <span class="cf">function</span>(path,bp){</span>
<span id="cb4-20"><a href="#cb4-20" tabindex="-1"></a>  d <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(path)</span>
<span id="cb4-21"><a href="#cb4-21" tabindex="-1"></a>  <span class="fu">approx</span>(d<span class="sc">$</span><span class="st">`</span><span class="at">Position(bp)</span><span class="st">`</span>, d<span class="sc">$</span><span class="st">`</span><span class="at">Map(cM)</span><span class="st">`</span>, <span class="at">xout =</span> bp, <span class="at">method =</span> <span class="st">&quot;linear&quot;</span>, <span class="at">rule =</span> <span class="dv">2</span>)<span class="sc">$</span>y</span>
<span id="cb4-22"><a href="#cb4-22" tabindex="-1"></a>}</span>
<span id="cb4-23"><a href="#cb4-23" tabindex="-1"></a></span>
<span id="cb4-24"><a href="#cb4-24" tabindex="-1"></a>dip2hapidx <span class="ot">&lt;-</span> <span class="cf">function</span>(x){</span>
<span id="cb4-25"><a href="#cb4-25" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="dv">2</span><span class="sc">*</span>x</span>
<span id="cb4-26"><a href="#cb4-26" tabindex="-1"></a>  <span class="fu">c</span>(<span class="fu">rbind</span>(x<span class="dv">-1</span>, x))</span>
<span id="cb4-27"><a href="#cb4-27" tabindex="-1"></a>}</span></code></pre></div>
<p>Now we can cluster <code>M</code> separately for African and
non-African haplotypes.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Load sample population information</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="do">#########################################</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>cluster_by <span class="ot">&lt;-</span> <span class="st">&quot;isAFR&quot;</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>id <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="fu">paste0</span>(data_dir, gene, <span class="st">&quot;.sample&quot;</span>))<span class="sc">$</span>sample</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>init_order_samples <span class="ot">&lt;-</span> <span class="fu">order</span>(id)</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">merge</span>(<span class="fu">data.table</span>(<span class="st">&quot;id&quot;</span> <span class="ot">=</span> id), kgp3, <span class="at">by =</span> <span class="st">&quot;id&quot;</span>)</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">nrow</span>(samples) <span class="sc">!=</span> <span class="fu">length</span>(id)) { <span class="fu">stop</span>(<span class="st">&quot;some samples have been removed by merging with kgp3&quot;</span>) }</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">all.equal</span>(init_order_samples, <span class="fu">order</span>(samples<span class="sc">$</span>id))) { <span class="fu">stop</span>(<span class="st">&quot;some samples have been moved out of the order in lct.sample&quot;</span>) }</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a>samples[,isAFR <span class="sc">:=</span> <span class="fu">ifelse</span>(reg <span class="sc">==</span> <span class="st">&quot;AFR&quot;</span>, <span class="st">&quot;AFR&quot;</span>, <span class="st">&quot;not_AFR&quot;</span>)]</span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="co"># Symmeterize &amp; Scale Distance Matrix at LCT</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="do">###############################################</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a>M <span class="ot">&lt;-</span> (<span class="fl">0.5</span><span class="sc">/</span>(neg_log10_mu<span class="sc">*</span><span class="fu">log</span>(<span class="dv">10</span>))) <span class="sc">*</span> (M <span class="sc">+</span> <span class="fu">t</span>(M))</span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a></span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a><span class="co"># Permute &amp; Cluster Distance Matrix</span></span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a><span class="do">###################################################################</span></span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a>diploid_perm <span class="ot">&lt;-</span> <span class="fu">order</span>(samples<span class="sc">$</span>reg, samples<span class="sc">$</span>pop, samples<span class="sc">$</span>id)</span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a>psamples <span class="ot">&lt;-</span> samples[diploid_perm,]</span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a></span>
<span id="cb5-20"><a href="#cb5-20" tabindex="-1"></a>haploid_perm <span class="ot">&lt;-</span> <span class="fu">dip2hapidx</span>(diploid_perm)</span>
<span id="cb5-21"><a href="#cb5-21" tabindex="-1"></a></span>
<span id="cb5-22"><a href="#cb5-22" tabindex="-1"></a>pM <span class="ot">&lt;-</span> M[, haploid_perm][haploid_perm,]</span>
<span id="cb5-23"><a href="#cb5-23" tabindex="-1"></a></span>
<span id="cb5-24"><a href="#cb5-24" tabindex="-1"></a>hap_groups <span class="ot">&lt;-</span> <span class="fu">table</span>(psamples[[cluster_by]])</span>
<span id="cb5-25"><a href="#cb5-25" tabindex="-1"></a>hap_groups <span class="ot">&lt;-</span> hap_groups[<span class="fu">unique</span>(psamples[[cluster_by]])]</span>
<span id="cb5-26"><a href="#cb5-26" tabindex="-1"></a></span>
<span id="cb5-27"><a href="#cb5-27" tabindex="-1"></a>baseline_idx <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fu">cumsum</span>(<span class="dv">2</span><span class="sc">*</span>hap_groups)[<span class="sc">-</span><span class="fu">length</span>(hap_groups)])</span>
<span id="cb5-28"><a href="#cb5-28" tabindex="-1"></a><span class="fu">names</span>(baseline_idx) <span class="ot">&lt;-</span> <span class="fu">names</span>(hap_groups)</span>
<span id="cb5-29"><a href="#cb5-29" tabindex="-1"></a></span>
<span id="cb5-30"><a href="#cb5-30" tabindex="-1"></a>order_M <span class="ot">&lt;-</span> <span class="fu">as.list</span>(hap_groups)</span>
<span id="cb5-31"><a href="#cb5-31" tabindex="-1"></a><span class="fu">names</span>(order_M) <span class="ot">&lt;-</span> <span class="fu">names</span>(hap_groups)</span>
<span id="cb5-32"><a href="#cb5-32" tabindex="-1"></a></span>
<span id="cb5-33"><a href="#cb5-33" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(hap_groups)){</span>
<span id="cb5-34"><a href="#cb5-34" tabindex="-1"></a>  current_pop_samples <span class="ot">&lt;-</span> <span class="fu">which</span>(psamples[[cluster_by]] <span class="sc">==</span> <span class="fu">names</span>(hap_groups)[i])</span>
<span id="cb5-35"><a href="#cb5-35" tabindex="-1"></a>  current_pop_haplotypes <span class="ot">&lt;-</span> <span class="fu">dip2hapidx</span>(current_pop_samples)</span>
<span id="cb5-36"><a href="#cb5-36" tabindex="-1"></a>  sM <span class="ot">&lt;-</span> pM[current_pop_haplotypes, current_pop_haplotypes]</span>
<span id="cb5-37"><a href="#cb5-37" tabindex="-1"></a>  order_M[[<span class="fu">names</span>(hap_groups)[i]]] <span class="ot">&lt;-</span> baseline_idx[<span class="fu">names</span>(hap_groups)[i]] <span class="sc">+</span> fastcluster<span class="sc">::</span><span class="fu">hclust</span>(<span class="fu">as.dist</span>(sM), <span class="at">method=</span><span class="st">&quot;average&quot;</span>)<span class="sc">$</span>order</span>
<span id="cb5-38"><a href="#cb5-38" tabindex="-1"></a>}</span>
<span id="cb5-39"><a href="#cb5-39" tabindex="-1"></a></span>
<span id="cb5-40"><a href="#cb5-40" tabindex="-1"></a>order_M <span class="ot">&lt;-</span> <span class="fu">unlist</span>(order_M)</span>
<span id="cb5-41"><a href="#cb5-41" tabindex="-1"></a>cM <span class="ot">&lt;-</span> pM[, order_M][order_M,]</span>
<span id="cb5-42"><a href="#cb5-42" tabindex="-1"></a></span>
<span id="cb5-43"><a href="#cb5-43" tabindex="-1"></a></span>
<span id="cb5-44"><a href="#cb5-44" tabindex="-1"></a><span class="co"># Plot clustered Distance Matrix </span></span>
<span id="cb5-45"><a href="#cb5-45" tabindex="-1"></a><span class="do">#########################################</span></span>
<span id="cb5-46"><a href="#cb5-46" tabindex="-1"></a><span class="fu">plot_mat</span>(cM, <span class="fu">paste0</span>(data_dir, gene, <span class="st">&quot;_dist_mat.pdf&quot;</span>))</span></code></pre></div>
</div>
<div id="reconstructing-lct-files-directly-from-1000-genomes-data-with-bcftools" class="section level1">
<h1>Reconstructing LCT files directly from 1000 Genomes data with
bcftools</h1>
<p>Download a phased VCF of Chromosome 2 from the 1000 Genomes website,
<code>https://www.internationalgenome.org/</code>. Below, we assume that
VCF has name <code>chr2.vcf.gz</code>. The 1000 Genomes Project
periodically makes updates to the VCFs available. We used
<code>ALLchr2phase3_shapeit2_mvncall_integrated_v5a20130502genotypes.vcf.gz</code>
in our analysis. That data was in hg19 coordinates, so throughout this
vignette, we work entirely in hg19 coordinates.</p>
<p>Please be mindful that some modification of the recombination map and
target locus index will be necessary if working with data based on a
different build.</p>
<p>If <code>bcftools</code> is not already installed, first install
<code>bcftools</code> from
<code>https://samtools.github.io/bcftools/</code>. Then from a bash
terminal, run</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">--regions</span> 2:136608646-136608646 <span class="at">--types</span> snps <span class="at">--min-ac</span> 2:minor <span class="at">-Ou</span> <span class="at">--threads</span> 1 chr2.vcf.gz <span class="kw">|</span> <span class="ex">bcftools</span> convert <span class="at">-h</span> lct <span class="at">--threads</span> 1</span></code></pre></div>
<p>Note, both of the above commands take a <code>--threads</code>
argument. Here, we’ve set it to <code>1</code> but this can increased to
the number of available cores to increase the execution speed. See the
<code>bcftools</code> documentation for more details.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
