<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Fast clade matrix construction — CladeMat • kalis</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/PT_Serif-0.4.10/font.css" rel="stylesheet"><link href="../deps/JetBrains_Mono-0.4.10/font.css" rel="stylesheet"><link href="../deps/Roboto_Slab-0.4.10/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Fast clade matrix construction — CladeMat"><meta name="description" content="Constructs a clade matrix using forward and backward tables.
The clade matrix captures genetic relatedness information in the distances from the Li &amp;amp; Stephens model that are not captured in the called clades."><meta property="og:description" content="Constructs a clade matrix using forward and backward tables.
The clade matrix captures genetic relatedness information in the distances from the Li &amp;amp; Stephens model that are not captured in the called clades."><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous"><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">kalis</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Decoding_a_Single_Variant.html">Decoding a Single Variant</a></li>
    <li><a class="dropdown-item" href="../articles/Iterating_Over_Loci.html">Iterating Over Loci</a></li>
    <li><a class="dropdown-item" href="../articles/lct_example.html">Reproducing kalis Paper LCT Example</a></li>
    <li><a class="dropdown-item" href="../articles/Reading_Haplotype_Data.html">Reading Haplotype Data</a></li>
  </ul></li>
<li class="nav-item"><a class="external-link nav-link" href="https://doi.org/10.1186/s12859-024-05688-8">BMC Bioinformatics Paper</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/louisaslett/kalis/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Fast clade matrix construction</h1>
      <small class="dont-index">Source: <a href="https://github.com/louisaslett/kalis/blob/master/R/CladeMat.R" class="external-link"><code>R/CladeMat.R</code></a></small>
      <div class="d-none name"><code>CladeMat.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Constructs a clade matrix using forward and backward tables.
The clade matrix captures genetic relatedness information in the distances from the Li &amp; Stephens model that are not captured in the called clades.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">CladeMat</span><span class="op">(</span></span>
<span>  <span class="va">fwd</span>,</span>
<span>  <span class="va">bck</span>,</span>
<span>  <span class="va">M</span>,</span>
<span>  <span class="va">unit.dist</span>,</span>
<span>  thresh <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  max1var <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  nthreads <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span>logical <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, <span class="va">fwd</span><span class="op">$</span><span class="va">to_recipient</span> <span class="op">-</span></span>
<span>    <span class="va">fwd</span><span class="op">$</span><span class="va">from_recipient</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-fwd">fwd<a class="anchor" aria-label="anchor" href="#arg-fwd"></a></dt>
<dd><p>a <code>kalisForwardTable</code> object, as returned by <code><a href="MakeForwardTable.html">MakeForwardTable()</a></code> and propagated to a target variant by <code><a href="Forward.html">Forward()</a></code>.
This table must be at the same variant location as argument <code>bck</code>.</p></dd>


<dt id="arg-bck">bck<a class="anchor" aria-label="anchor" href="#arg-bck"></a></dt>
<dd><p>a <code>kalisBackwardTable</code> object, as returned by <code><a href="MakeBackwardTable.html">MakeBackwardTable()</a></code> and propagated to a target variant by <code><a href="Backward.html">Backward()</a></code>.
This table must be at the same variant location as argument <code>fwd</code>.</p></dd>


<dt id="arg-m">M<a class="anchor" aria-label="anchor" href="#arg-m"></a></dt>
<dd><p>a matrix with half the number of rows and columns as the corresponding forward/backward tables.
This matrix is overwritten in place with the clade matrix result for performance reasons.</p></dd>


<dt id="arg-unit-dist">unit.dist<a class="anchor" aria-label="anchor" href="#arg-unit-dist"></a></dt>
<dd><p>the change in distance that is expected to correspond to a single mutation (typically \(-\log(\mu)\)) for the LS model)</p></dd>


<dt id="arg-thresh">thresh<a class="anchor" aria-label="anchor" href="#arg-thresh"></a></dt>
<dd><p>a regularization parameter: differences of distances must exceed this threshold (in <code>unit.dist</code> units) in order to cause the introduction of a probabilistic clade.
Defaults to <code>0.2</code>.</p></dd>


<dt id="arg-max-var">max1var<a class="anchor" aria-label="anchor" href="#arg-max-var"></a></dt>
<dd><p>a logical regularization parameter.
When <code>TRUE</code>, differences in distances exceeding 1 <code>unit.dist</code> are set to 1 (so that any edge in the latent ancestral tree with multiple mutations on them are treated as if only one mutation was on it).</p></dd>


<dt id="arg-nthreads">nthreads<a class="anchor" aria-label="anchor" href="#arg-nthreads"></a></dt>
<dd><p>the number of CPU cores to use.
By default uses the <code>parallel</code> package to detect the number of physical cores.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list, the first element contains a list of tied nearest neighbours (one for each haplotype).
Other elements of the returned list are for internal use by <code><a href="PruneCladeMat.html">PruneCladeMat()</a></code> to allow for efficient removal of singletons and sprigs.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><code>CladeMat()</code> uses the forward and backward tables to construct the corresponding clade matrix which can then be tested, for example using a standard quadratic form score statistic.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Christ, R.R., Wang, X., Aslett, L.J.M., Steinsaltz, D. and Hall, I. (2024) "Clade Distillation for Genome-wide Association Studies", bioRxiv 2024.09.30.615852. Available at: <a href="https://doi.org/10.1101/2024.09.30.615852" class="external-link">doi:10.1101/2024.09.30.615852</a>
.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># TODO</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Louis Aslett, Ryan Christ.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

