<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Create a kalisForwardTable — MakeForwardTable • kalis</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/PT_Serif-0.4.9/font.css" rel="stylesheet"><link href="../deps/JetBrains_Mono-0.4.9/font.css" rel="stylesheet"><link href="../deps/Roboto_Slab-0.4.9/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Create a kalisForwardTable — MakeForwardTable"><meta name="description" content="Allocates the memory for and initializes a forward table."><meta property="og:description" content="Allocates the memory for and initializes a forward table."><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous"><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">kalis</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Decoding_a_Single_Variant.html">Decoding a Single Variant</a></li>
    <li><a class="dropdown-item" href="../articles/Iterating_Over_Loci.html">Iterating Over Loci</a></li>
    <li><a class="dropdown-item" href="../articles/lct_example.html">Reproducing kalis Paper LCT Example</a></li>
    <li><a class="dropdown-item" href="../articles/Reading_Haplotype_Data.html">Reading Haplotype Data</a></li>
  </ul></li>
<li class="nav-item"><a class="external-link nav-link" href="https://doi.org/10.1186/s12859-024-05688-8">BMC Bioinformatics Paper</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/louisaslett/kalis/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Create a <code>kalisForwardTable</code></h1>
      <small class="dont-index">Source: <a href="https://github.com/louisaslett/kalis/blob/master/R/TableMaker.R" class="external-link"><code>R/TableMaker.R</code></a></small>
      <div class="d-none name"><code>MakeForwardTable.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Allocates the memory for and initializes a forward table.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">MakeForwardTable</span><span class="op">(</span><span class="va">pars</span>, from_recipient <span class="op">=</span> <span class="fl">1</span>, to_recipient <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-pars">pars<a class="anchor" aria-label="anchor" href="#arg-pars"></a></dt>
<dd><p>a <code>kalisParameters</code> object specifying the genetics parameters to be associated with this forward table.
These parameters can be set up by using the <code><a href="Parameters.html">Parameters()</a></code> function.</p></dd>


<dt id="arg-from-recipient">from_recipient<a class="anchor" aria-label="anchor" href="#arg-from-recipient"></a></dt>
<dd><p>first recipient haplotype included if creating a partial forward table.
By default all are included from the first recipient haplotype.
Haplotypes are indexed from 1.</p></dd>


<dt id="arg-to-recipient">to_recipient<a class="anchor" aria-label="anchor" href="#arg-to-recipient"></a></dt>
<dd><p>last recipient haplotype included if creating a partial forward table.
By default all are included upto the last recipient haplotype.
Haplotypes are indexed from 1.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A specialized list of class <code>kalisForwardTable</code>.
The elements of the forward table list are:</p><dl><dt><code>l</code></dt>
<dd><p>denotes the current variant position.
2147483647 indicates a newly created forward table which has not yet been propagated to any variant.</p></dd>

<dt><code>alpha</code></dt>
<dd><p>is a matrix of rescaled forward probabilities under the Li and Stephens HMM.
Each column of <code>alpha</code> corresponds to an independent HMM such that \(\alpha^l_{ji}\) is proportional to the probability that haplotype \(j\) is copied by haplotype \(i\) at variant \(l\) and observing haplotype \(i\) from variant 1 up through variant \(l\).</p></dd>

<dt><code>alpha.f</code></dt>
<dd><p>is a vector containing scaling constants needed to continue propagating the HMM (TODO: add ref please see kalis paper for details).</p></dd>


</dl><p>A <code>kalisForwardTable</code> also carries with it a checksum key for the parameters <code>pars</code> it was provided.
If one attempts to interact a <code>kalisForwardTable</code> with a <code>kalisBackwardTable</code> with mismatched parameters, an error will be thrown.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><code>MakeForwardTable</code> initializes a <code>kalisForwardTable</code> object that will store the rescaled forward probabilities for each recipient haplotype.
Note: since all other haplotypes loaded by <code><a href="CacheHaplotypes.html">CacheHaplotypes()</a></code> into the kalis cache are taken as potential donor haplotypes, this table will be of size (total haplotypes in cache)x(number of recipients).
Also note that this table object is linked to a given set of Li and Stephens hidden Markov model parameters, created by <code><a href="Parameters.html">Parameters()</a></code>, to ensure consistency for any given HMM run.</p>
<p>The returned <code>kalisForwardTable</code> is ready to be propagated to a given target variant with the function <code><a href="Forward.html">Forward()</a></code>.</p>
<p>Since each column corresponds to an independent Li and Stephens hidden Markov model (ie for each recipient), it is possible to create a partial forward table object which corresponds to a subset of recipients using the <code>from_recipient</code> and <code>to_recipient</code> arguments.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Aslett, L.J.M. and Christ, R.R. (2024) "kalis: a modern implementation of the Li &amp; Stephens model for local ancestry inference in R", <em>BMC Bioinformatics</em>, <strong>25</strong>(1). Available at: <a href="https://doi.org/10.1186/s12859-024-05688-8" class="external-link">doi:10.1186/s12859-024-05688-8</a>
.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="Forward.html">Forward()</a></code> to propagate the newly created <code>kalisForwardTable</code>;
<code><a href="MakeBackwardTable.html">MakeBackwardTable()</a></code> to create a corresponding <code>kalisBackwardTable</code>.</p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Load the toy haplotype example and set toy parameters</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"SmallHaps"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"SmallMap"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="CacheHaplotypes.html">CacheHaplotypes</a></span><span class="op">(</span><span class="va">SmallHaps</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>haplotypes already cached ... overwriting existing cache.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">rho</span> <span class="op">&lt;-</span> <span class="fu"><a href="CalcRho.html">CalcRho</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="va">SmallMap</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">pars</span> <span class="op">&lt;-</span> <span class="fu"><a href="Parameters.html">Parameters</a></span><span class="op">(</span><span class="va">rho</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create the forward table we want to propagate</span></span></span>
<span class="r-in"><span><span class="va">fwd</span> <span class="op">&lt;-</span> <span class="fu">MakeForwardTable</span><span class="op">(</span><span class="va">pars</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fwd</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Full Forward Table object for 300 haplotypes. </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Newly created table, currently uninitialised to any variant (ready for Forward function next).</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Memory consumed: 723.98 kB </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Now ready to run the HMM forward recursions with Forward() ...</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Louis Aslett, Ryan Christ.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

