<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Posterior marginal probabilities — PostProbs • kalis</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/PT_Serif-0.4.10/font.css" rel="stylesheet"><link href="../deps/JetBrains_Mono-0.4.10/font.css" rel="stylesheet"><link href="../deps/Roboto_Slab-0.4.10/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Posterior marginal probabilities — PostProbs"><meta name="description" content="Calculate the posterior marginal probabilities at a given variant using forward and backward tables propagated to that position."><meta property="og:description" content="Calculate the posterior marginal probabilities at a given variant using forward and backward tables propagated to that position."><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous"><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">kalis</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Decoding_a_Single_Variant.html">Decoding a Single Variant</a></li>
    <li><a class="dropdown-item" href="../articles/Iterating_Over_Loci.html">Iterating Over Loci</a></li>
    <li><a class="dropdown-item" href="../articles/lct_example.html">Reproducing kalis Paper LCT Example</a></li>
    <li><a class="dropdown-item" href="../articles/Reading_Haplotype_Data.html">Reading Haplotype Data</a></li>
  </ul></li>
<li class="nav-item"><a class="external-link nav-link" href="https://doi.org/10.1186/s12859-024-05688-8">BMC Bioinformatics Paper</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/louisaslett/kalis/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Posterior marginal probabilities</h1>
      <small class="dont-index">Source: <a href="https://github.com/louisaslett/kalis/blob/master/R/Probs.R" class="external-link"><code>R/Probs.R</code></a></small>
      <div class="d-none name"><code>PostProbs.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Calculate the posterior marginal probabilities at a given variant using forward and backward tables propagated to that position.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">PostProbs</span><span class="op">(</span></span>
<span>  <span class="va">fwd</span>,</span>
<span>  <span class="va">bck</span>,</span>
<span>  unif.on.underflow <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  M <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  beta.theta.opts <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nthreads <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span>logical <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, <span class="va">fwd</span><span class="op">$</span><span class="va">to_recipient</span> <span class="op">-</span></span>
<span>    <span class="va">fwd</span><span class="op">$</span><span class="va">from_recipient</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-fwd">fwd<a class="anchor" aria-label="anchor" href="#arg-fwd"></a></dt>
<dd><p>a forward table as returned by <code><a href="MakeForwardTable.html">MakeForwardTable()</a></code> and propagated to a target variant by <code><a href="Forward.html">Forward()</a></code>.
Must be at the same variant as <code>bck</code> (unless <code>bck</code> is in "beta-theta space" in which case if must be downstream ... see <code><a href="Backward.html">Backward()</a></code> for details).</p></dd>


<dt id="arg-bck">bck<a class="anchor" aria-label="anchor" href="#arg-bck"></a></dt>
<dd><p>a backward table as returned by <code><a href="MakeBackwardTable.html">MakeBackwardTable()</a></code> and propagated to a target variant by <code><a href="Backward.html">Backward()</a></code>.
Must be at the same variant as <code>fwd</code> (unless <code>bck</code> is in "beta-theta space" in which case if must be downstream ... see <code><a href="Backward.html">Backward()</a></code> for details).</p></dd>


<dt id="arg-unif-on-underflow">unif.on.underflow<a class="anchor" aria-label="anchor" href="#arg-unif-on-underflow"></a></dt>
<dd><p>a logical; if <code>TRUE</code>, then if all probabilities in a column underflow, then they will be set to \(\frac{1}{N-1}\) instead of 0</p></dd>


<dt id="arg-m">M<a class="anchor" aria-label="anchor" href="#arg-m"></a></dt>
<dd><p>a pre-existing matrix into which to write the probabilities, can yield substantial speed up but requires special attention (see Details)</p></dd>


<dt id="arg-beta-theta-opts">beta.theta.opts<a class="anchor" aria-label="anchor" href="#arg-beta-theta-opts"></a></dt>
<dd><p>a list; see Details.</p></dd>


<dt id="arg-nthreads">nthreads<a class="anchor" aria-label="anchor" href="#arg-nthreads"></a></dt>
<dd><p>the number of CPU cores to use.
By default uses the <code>parallel</code> package to detect the number of physical cores.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Matrix of posterior marginal probabilities.
The \((j,i)\)-th element of of the returned matrix is the probability that \(j\) is copied by \(i\) at the current variant, \(l\), of the two tables, given the haplotypes observed (over the whole sequence).</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The forward and backward tables must be at the same variant in order for them to be combined to yield the posterior marginal probabilities at variant \(l\).
The \((j,i)\)-th element of of the returned matrix is the probability that \(j\) is copied by \(i\) at the current variant, \(l\), of the two tables, given the haplotypes observed (over the whole sequence).</p>
<p>Note that each column represents an independent HMM.</p>
<p>By convention, every diagonal element is zero.</p>
<p><strong>Notes on <code>beta.theta.opts</code></strong></p>
<p>In order to obtain posterior marginal probability matrices between variants <code>fwd$l</code> and <code>bck$l</code>, then <code>bck</code> must be in "beta-theta space", see <code><a href="Backward.html">Backward()</a></code> for details.
This allows the forward and backward tables to be transitioning both tables to some genomic position between <code>fwd$l</code> and <code>bck$l</code>.
The precise recombination distance by which each table is propagated can be controlled by passing optional arguments in a list via <code>beta.theta.opts</code>.
The recombination distances used can be specified in one of two ways.</p><ol><li><p>Manually.
In this case, <code>beta.theta.opts</code> is a list containing two named elements:</p><ul><li><p><code>"rho.fwd"</code> \(\in (0,1)\) specifies the transition probability <code>rho</code> for propagating the forward table.</p></li>
<li><p><code>"rho.bck"</code> \(\in (0,1)\) specifies the transition probability <code>rho</code> for propagating the backward table.</p></li>
</ul></li>
<li><p>Implicitly.
In this case, <code>beta.theta.opts</code> is a list containing two named elements:</p></li>
</ol><ul><li><p><code>"pars"</code>: a <code>kalisParameters</code> object that implicitly defines the recombination distance \(\rho^\star\) between <code>fwd$l</code> and <code>bck$l</code></p></li>
<li><p><code>"bias"</code> \(\in (0,1)\).
The forward table is propagated a distance of <code>(bias)</code>\(\rho^\star\) and the backward table is propagated a distance of <code>(1-bias)</code>\(\rho^\star\).</p></li>
</ul><p><strong>Performance notes</strong></p>
<p>If calculating many posterior probability matrices in succession, providing a pre-existing matrix <code>M</code> that can be updated in-place can dramatically increase speed by eliminating the time needed for memory allocation.
Be warned, since the matrix is updated in-place, if any other variables point to the same memory address, they will also be simultaneously overwritten.
For example, writing</p>
<p></p><div class="sourceCode"><pre><code><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a>M <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">0</span>, <span class="fu">nrow</span>(fwd<span class="sc">$</span>alpha), <span class="fu">ncol</span>(fwd<span class="sc">$</span>alpha))</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>P <span class="ot">&lt;-</span> M</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">PostProbs</span>(fwd, bck, <span class="at">M =</span> M)</span></code></pre><p></p></div>
<p>will update <code>M</code> and <code>P</code> simultaneously.</p>
<p>When provided, <code>M</code> must have dimensions matching that of <code>fwd$alpha</code>.
Typically, that is simply \(N \times N\) for \(N\) haplotypes.
However, if kalis is being run in a distributed manner, <code>M</code> will be a \(N \times R\) matrix where \(R\) is the number of recipient haplotypes on the current machine.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Aslett, L.J.M. and Christ, R.R. (2024) "kalis: a modern implementation of the Li &amp; Stephens model for local ancestry inference in R", <em>BMC Bioinformatics</em>, <strong>25</strong>(1). Available at: <a href="https://doi.org/10.1186/s12859-024-05688-8" class="external-link">doi:10.1186/s12859-024-05688-8</a>
.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="DistMat.html">DistMat()</a></code> to generate calculate \(d_{ji}\) distances directly;
<code><a href="Forward.html">Forward()</a></code> to propagate a Forward table to a new variant;
<code><a href="Backward.html">Backward()</a></code> to propagate a Backward table to a new variant</p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># To get the posterior probabilities at, say, variants 100 of the toy data</span></span></span>
<span class="r-in"><span><span class="co"># built into kalis</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"SmallHaps"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"SmallMap"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="CacheHaplotypes.html">CacheHaplotypes</a></span><span class="op">(</span><span class="va">SmallHaps</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>haplotypes already cached ... overwriting existing cache.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">rho</span> <span class="op">&lt;-</span> <span class="fu"><a href="CalcRho.html">CalcRho</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="va">SmallMap</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">pars</span> <span class="op">&lt;-</span> <span class="fu"><a href="Parameters.html">Parameters</a></span><span class="op">(</span><span class="va">rho</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">fwd</span> <span class="op">&lt;-</span> <span class="fu"><a href="MakeForwardTable.html">MakeForwardTable</a></span><span class="op">(</span><span class="va">pars</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bck</span> <span class="op">&lt;-</span> <span class="fu"><a href="MakeBackwardTable.html">MakeBackwardTable</a></span><span class="op">(</span><span class="va">pars</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="Forward.html">Forward</a></span><span class="op">(</span><span class="va">fwd</span>, <span class="va">pars</span>, <span class="fl">100</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="Backward.html">Backward</a></span><span class="op">(</span><span class="va">bck</span>, <span class="va">pars</span>, <span class="fl">100</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">PostProbs</span><span class="op">(</span><span class="va">fwd</span>, <span class="va">bck</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="DistMat.html">DistMat</a></span><span class="op">(</span><span class="va">fwd</span>, <span class="va">bck</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Louis Aslett, Ryan Christ.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

