<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Reads/writes an R matrix of 0/1s to the HDF5 format which is used for reading to the kalis optimised memory cache.
If you're working with a large haplotype dataset, we recommend that you convert it directly to this HDF5 format (see vignette) rather than read it into R."><title>I/O for haplotype matrices in HDF5 files — WriteHaplotypes • kalis</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/PT_Serif-0.4.8/font.css" rel="stylesheet"><link href="../deps/JetBrains_Mono-0.4.8/font.css" rel="stylesheet"><link href="../deps/Roboto_Slab-0.4.8/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="I/O for haplotype matrices in HDF5 files — WriteHaplotypes"><meta property="og:description" content="Reads/writes an R matrix of 0/1s to the HDF5 format which is used for reading to the kalis optimised memory cache.
If you're working with a large haplotype dataset, we recommend that you convert it directly to this HDF5 format (see vignette) rather than read it into R."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">kalis</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Decoding_a_Single_Variant.html">Decoding a Single Variant</a>
    <a class="dropdown-item" href="../articles/lct_example.html">Reproducing kalis Paper LCT Example</a>
  </div>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://doi.org/10.1186/s12859-024-05688-8">BMC Bioinformatics Paper</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/louisaslett/kalis/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>I/O for haplotype matrices in HDF5 files</h1>
      <small class="dont-index">Source: <a href="https://github.com/louisaslett/kalis/blob/HEAD/R/IndividualSequenceIO_H5.R" class="external-link"><code>R/IndividualSequenceIO_H5.R</code></a></small>
      <div class="d-none name"><code>WriteHaplotypes.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Reads/writes an R matrix of 0/1s to the HDF5 format which is used for reading to the kalis optimised memory cache.
If you're working with a large haplotype dataset, we recommend that you convert it directly to this HDF5 format (see vignette) rather than read it into R.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">WriteHaplotypes</span><span class="op">(</span></span>
<span>  <span class="va">hdf5.file</span>,</span>
<span>  <span class="va">haps</span>,</span>
<span>  hap.ids <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  loci.ids <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  haps.name <span class="op">=</span> <span class="st">"/haps"</span>,</span>
<span>  hap.ids.name <span class="op">=</span> <span class="st">"/hap.ids"</span>,</span>
<span>  loci.ids.name <span class="op">=</span> <span class="st">"/loci.ids"</span>,</span>
<span>  append <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ReadHaplotypes</span><span class="op">(</span></span>
<span>  <span class="va">hdf5.file</span>,</span>
<span>  loci.idx <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  hap.idx <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  loci.ids <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  hap.ids <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  haps.name <span class="op">=</span> <span class="st">"/haps"</span>,</span>
<span>  loci.ids.name <span class="op">=</span> <span class="st">"/loci.ids"</span>,</span>
<span>  hap.ids.name <span class="op">=</span> <span class="st">"/hap.ids"</span>,</span>
<span>  transpose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>hdf5.file</dt>
<dd><p>the name of the file which the haplotypes are to be written to.</p></dd>


<dt>haps</dt>
<dd><p>a vector or a matrix where each column is a haplotype to be stored in the file <code>hdf5.file</code>.</p></dd>


<dt>hap.ids</dt>
<dd><p>a character vector naming haplotypes when writing, or which haplotypes are to be read.</p></dd>


<dt>loci.ids</dt>
<dd><p>a character vector naming variants when writing, or which variants are to be read.</p></dd>


<dt>haps.name</dt>
<dd><p>a string providing the full path and object name where the haplotype matrix should be read/written.</p></dd>


<dt>hap.ids.name</dt>
<dd><p>a string providing the full path and object name where the haplotype names (in <code>haps.ids</code>) should be read/written.</p></dd>


<dt>loci.ids.name</dt>
<dd><p>a string providing the full path and object name where the variant names (in <code>loci.ids</code>) should be read/written.</p></dd>


<dt>append</dt>
<dd><p>a logical indicating whether overwrite (default) or append to an existing <code>haps</code> dataset if it already exists in <code>hdf5.file</code>.</p></dd>


<dt>loci.idx</dt>
<dd><p>an integer vector of the indices of which variants are to be read (for naming, use <code>hap.ids</code>).</p></dd>


<dt>hap.idx</dt>
<dd><p>an integer vector of the indices of which haplotypes are to be read (for naming, use <code>hap.ids</code>).</p></dd>


<dt>transpose</dt>
<dd><p>a logical indicating whether to transpose the logic of haplotypes/variants when reading.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p><code>WriteHaplotypes</code> does not return anything.</p>


<p><code>ReadHaplotypes</code> returns a binary matrix containing the
haplotypes that were specified in <code>ids</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The primary method to load data into kalis' internal optimised cache is from an HDF5 storage file.
If the user has a collection of haplotypes already represented as a matrix of 0's and 1's in R, this function can be used to write to HDF5 in the format required to load into cache.</p>
<p>kalis expects a 2-dimensional object named <code>haps</code> at the root level of the HDF5 file.
Haplotypes should be stored in the slowest changing dimension as defined in the HDF5 specification (note that different languages treat this as rows or columns).</p>
<p>Note that if <code>hdf5.file</code> exists but does not contain a dataset named <code>haps</code>, then <code>WriteHaplotypes</code> will simply create a <code>haps</code> dataset within the existing file.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="CacheHaplotypes.html">CacheHaplotypes()</a></code> to fill the kalis cache with haplotypes.</p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="co"># Generate a random mini set of haplotypes to write</span></span></span>
<span class="r-in"><span><span class="va">n.haps</span> <span class="op">&lt;-</span> <span class="fl">20</span></span></span>
<span class="r-in"><span><span class="va">n.vars</span> <span class="op">&lt;-</span> <span class="fl">200</span></span></span>
<span class="r-in"><span><span class="va">haps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">1</span>, <span class="va">n.haps</span><span class="op">*</span><span class="va">n.vars</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>               nrow <span class="op">=</span> <span class="va">n.vars</span>, ncol <span class="op">=</span> <span class="va">n.haps</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># ... write them to a file, giving alphabetic letters "A" through "T" as the #' # haplotype names ...</span></span></span>
<span class="r-in"><span><span class="fu">WriteHaplotypes</span><span class="op">(</span><span class="st">"~/myhaps.h5"</span>, <span class="va">haps</span>, hap.ids <span class="op">=</span> <span class="va">LETTERS</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating HDF5 file ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Writing 20 haplotype(s) of size 200 ...</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># ... and confirm we can read a chosen portion back.  Try to read back</span></span></span>
<span class="r-in"><span><span class="co"># the 10th and 11th haplotypes by using their name (J and K are 10th and 11th</span></span></span>
<span class="r-in"><span><span class="co"># letter of the alphabet)</span></span></span>
<span class="r-in"><span><span class="va">h5</span> <span class="op">&lt;-</span> <span class="fu">ReadHaplotypes</span><span class="op">(</span><span class="st">"~/myhaps.h5"</span>, hap.ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"J"</span>,<span class="st">"K"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="va">h5</span><span class="op">$</span><span class="va">haps</span> <span class="op">==</span> <span class="va">haps</span><span class="op">[</span>, <span class="fl">10</span><span class="op">:</span><span class="fl">11</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Read from the .h5 file into the kalis cache and confirm that what we wrote</span></span></span>
<span class="r-in"><span><span class="co"># out to the HDF5 file matches the original matrix we generated in R</span></span></span>
<span class="r-in"><span><span class="fu"><a href="CacheHaplotypes.html">CacheHaplotypes</a></span><span class="op">(</span><span class="st">"~/myhaps.h5"</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>haplotypes already cached ... overwriting existing cache.</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="va">haps</span> <span class="op">==</span> <span class="fu"><a href="QueryCache.html">QueryCache</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-in"><span><span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Louis Aslett, Ryan Christ.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

