<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Distance matrix — DistMat • kalis</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Distance matrix — DistMat" />
<meta property="og:description" content="Utility for calculating posterior marginal probabilities and distance matrices at, in between, or excluding variants." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">kalis</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/AutoTuning.html">Auto-tuning HMM Parameters for Local Phylogeny Inference</a>
    </li>
    <li>
      <a href="../articles/Decoding_a_Single_Locus.html">Decoding a Single Locus</a>
    </li>
    <li>
      <a href="../articles/SimpleGWAS.html">SimpleGWAS</a>
    </li>
    <li>
      <a href="../articles/testing_new_dedip.html">testing_new_dedip</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Distance matrix</h1>
    
    <div class="hidden name"><code>DistMat.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Utility for calculating posterior marginal probabilities and distance matrices at, in between, or excluding variants.</p>
    </div>

    <pre class="usage"><span class='fu'>DistMat</span><span class='op'>(</span>
  <span class='va'>fwd</span>,
  <span class='va'>bck</span>,
  standardize <span class='op'>=</span> <span class='cn'>FALSE</span>,
  M <span class='op'>=</span> <span class='cn'>NULL</span>,
  beta.theta.opts <span class='op'>=</span> <span class='cn'>NULL</span>,
  nthreads <span class='op'>=</span> <span class='fl'>1</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>fwd</th>
      <td><p>a forward table as returned by <code><a href='MakeForwardTable.html'>MakeForwardTable</a></code></p></td>
    </tr>
    <tr>
      <th>bck</th>
      <td><p>a backward table as returned by <code><a href='MakeBackwardTable.html'>MakeBackwardTable</a></code></p></td>
    </tr>
    <tr>
      <th>standardize</th>
      <td><p>a logical; should the columns be centered and scaled to have unit variance</p></td>
    </tr>
    <tr>
      <th>M</th>
      <td><p>an pre-existing matrix into which to write the distances, can yield substantial speed up but requires special attention (see Details)</p></td>
    </tr>
    <tr>
      <th>beta.theta.opts</th>
      <td><p>a list; see Details.</p></td>
    </tr>
    <tr>
      <th>nthreads</th>
      <td><p>the number of CPU cores to use. By default no parallelism is used.</p></td>
    </tr>
    <tr>
      <th>from_recipient</th>
      <td><p>offset for distributed problems, see kalis distributed.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Matrix of distances.
The \((j,i)\)-th element of of the returned matrix is the inferred
distance \(d_{ji}\) to haplotype \(j\) from haplotype \(i\) at
the current locus.  Each column encodes the output of an independent HMM: in column \(i\), haplotype \(i\) is taken as the observed recipient haplotype and painted as a mosaic of the other \(N-1\) haplotypes.
Hence, the distances are asymmetric.</p>
<p>If calculating many distance matrices in succession, providing a pre-existing matrix <code>M</code> that can be updated in-place can drastically increase speed by eliminating the time needed for memory allocation.
Be warned, since the matrix is updated in-place, if any other variables point to the same memory address, they will also be simultaneously overwritten.  For example, writing</p><pre><span class='va'>M</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>matrix</a></span><span class='op'>(</span><span class='fl'>0</span>,<span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>fwd</span><span class='op'>$</span><span class='va'>alpha</span><span class='op'>)</span>,<span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>ncol</a></span><span class='op'>(</span><span class='va'>fwd</span><span class='op'>$</span><span class='va'>alpha</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>P</span> <span class='op'>&lt;-</span> <span class='va'>M</span>
<span class='fu'>DistMat</span><span class='op'>(</span><span class='va'>fwd</span>,<span class='va'>bck</span>,M<span class='op'>=</span><span class='va'>M</span><span class='op'>)</span>
</pre>

<p>will update M and P simultaneously.</p>
<p>When provided, <code>M</code> must have dimensions matching that of <code>fwd$alpha</code>.  Typically, that is simply \(N\) x \(N\) for \(N\) haplotypes.
However, if kalis is being run distributed, <code>M</code> will be a \(N\) x \(R\) matrix where \(R\) is the number of recipient haplotypes on the given machine.</p>
<p>In order to obtain distance matrices between variants <code>fwd$l</code> and <code>bck$l</code>, then <code>bck</code> must be in <code>beta.theta</code> space.  This allows the forward and backward tables to be transitioning both tables to some genomic position between <code>fwd$l</code> and <code>bck$l</code>.
The precise recombination distance by which each table is propagated can be controlled by passing optional arguments in a list via <code>beta.theta.opts</code>.
The recombination distances used can be specified in one of two ways.</p><ol>
<li><p>Manually.  In this case, <code>beta.theta.opts</code> is a list containing two named elements:</p><ul>
<li><p><code>"rho.fwd"</code> \(\in (0,1)\) specifies the transition probability \(rho\) for propagating the forward table</p></li>
<li><p><code>"rho.bck"</code> \(\in (0,1)\) specifies the transition probability \(rho\) for propagating the backward table.</p></li>
</ul></li>
<li><p>Implicitly.  In this case, <code>beta.theta.opts</code> is a list containing two named elements:</p><ul>
<li><p><code>"pars"</code>: a <code>kalisParameters</code> object that implicitly defines the recombination distance \(\rho^\star\) between <code>fwd$l</code> and <code>bck$l</code></p></li>
<li><p><code>"bias"</code> \(\in (0,1)\).  The forward table is propagated a distance of <code>bias</code>\(\rho^\star\) and the backward table is propagated a distance of <code>(1-bias)</code>\(\rho^\star\).</p></li>
</ul></li>
</ol>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This computes a local probability or distance matrix based on the forward and backward
tables at a certain locus.  The default usage is provide forward and backward tables at the same locus \(l\).
so that the \((j,i)\)-th element of of the returned matrix is the inferred distance
\(d_{ji}\) between haplotypes \(j\) and \(i\) at the current
locus, \(l\), of the two tables given the haplotypes observed (over the
whole sequence).</p>
<p>In particular,</p>
<p>$$d_{ji} = -log(p_{ji})$$</p>
<p>where \(p_{ji}\) is the posterior marginal probability that
\(j\) is coped by \(i\) at the current locus of the two tables, \(l\),
given the haplotypes observed (over the whole sequence).</p>
<p>By convention, \(d_{ii} = 0\) for all \(i\).</p>
<p>This function also allows users to calculate distance matrices in between variants and also to calculate matrices that exclude a set of consecutive variants by passing a
backward table in beta.theta space.  If in beta.theta space, <code>bck$l</code> may be greater than but not equal to <code>fwd$l</code>.  <code>beta.theta.opts</code> provides is required in this case to
set how much of a recombination distance to propagate each matrix before combining them into distances.  See Details below.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='PostProbs.html'>PostProbs</a></code> to calculate the posterior marginal probabilities
\(p_{ji}\);
<code><a href='Forward.html'>Forward</a></code> to propogate a Forward table to a new locus;
<code><a href='Backward.html'>Backward</a></code> to propogate a Backward table to a new locus.</p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='co'># To get the distances at, say, locus 10 ...</span>
<span class='fu'><a href='Forward.html'>Forward</a></span><span class='op'>(</span><span class='va'>fwd</span>, <span class='va'>pars</span>, <span class='fl'>10</span>, nthreads <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span>
<span class='fu'><a href='Backward.html'>Backward</a></span><span class='op'>(</span><span class='va'>bck</span>, <span class='va'>pars</span>, <span class='fl'>10</span>, nthreads <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span>
<span class='va'>d</span> <span class='op'>&lt;-</span> <span class='fu'>DistMat</span><span class='op'>(</span><span class='va'>fwd</span>, <span class='va'>bck</span><span class='op'>)</span>
<span class='op'>}</span>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Louis Aslett, Ryan Christ.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


