<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Decoding a Single Variant • kalis</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/PT_Serif-0.4.9/font.css" rel="stylesheet">
<link href="../deps/JetBrains_Mono-0.4.9/font.css" rel="stylesheet">
<link href="../deps/Roboto_Slab-0.4.9/font.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Decoding a Single Variant">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous">
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">kalis</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Decoding_a_Single_Variant.html">Decoding a Single Variant</a></li>
    <li><a class="dropdown-item" href="../articles/Iterating_Over_Loci.html">Iterating Over Loci</a></li>
    <li><a class="dropdown-item" href="../articles/lct_example.html">Reproducing kalis Paper LCT Example</a></li>
    <li><a class="dropdown-item" href="../articles/Reading_Haplotype_Data.html">Reading Haplotype Data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="external-link nav-link" href="https://doi.org/10.1186/s12859-024-05688-8">BMC Bioinformatics Paper</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/louisaslett/kalis/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Decoding a Single Variant</h1>
                        <h4 data-toc-skip class="author">Louis Aslett
&amp; Ryan Christ</h4>
            
            <h4 data-toc-skip class="date">2024-11-13</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/louisaslett/kalis/blob/master/vignettes/Decoding_a_Single_Variant.Rmd" class="external-link"><code>vignettes/Decoding_a_Single_Variant.Rmd</code></a></small>
      <div class="d-none name"><code>Decoding_a_Single_Variant.Rmd</code></div>
    </div>

    
    
<p>In order to demonstrate how to use to calculate and plot a local
distance matrix
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>d</mi><mo>ℓ</mo></msup><annotation encoding="application/x-tex">d^{\ell}</annotation></semantics></math>,
the package comes with an toy dataset of 300 simulated haplotypes, ,
stored as a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>L</mi><mo>=</mo><mn>400</mn></mrow><annotation encoding="application/x-tex">L = 400</annotation></semantics></math>
by
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>N</mi><mo>=</mo><mn>300</mn></mrow><annotation encoding="application/x-tex">N = 300</annotation></semantics></math>
.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://kalis.louisaslett.com/">kalis</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: kalis</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running in 64-bit mode using x86-64 architecture.</span></span>
<span><span class="co">#&gt; Loops unrolled to depth 4.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Currently not using any special instruction sets (WARNING: poor performance likely).</span></span>
<span><span class="co">#&gt; If this is unexpected (e.g. your CPU is Intel Haswell or newer architecture, or ARMv7+ with NEON support), then ensure that you are targeting the native architecture in compilation.  The easiest method is to add/change the following line in ~/.R/Makevars</span></span>
<span><span class="co">#&gt; CFLAGS=-march=native -mtune=native -O3</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"SmallHaps"</span><span class="op">)</span></span></code></pre></div>
<p> maintains a special cache for loading a haplotype dataset (see next
section for details). Since is designed for much larger datasets than
this toy example that we would often never want load into , haplotypes
can be imported directly into the cache using . can only cache and
operate on one haplotype dataset at a time. Currently, requires that the
haplotypes be stored in our custom HDF5 format but support for other
common formats will be added shortly. In the following section we
provide simple instructions for converting common file formats like VCFs
and HAP/SAMPLE/LEGEND generated by SHAPEIT2 to our HDF5 format. However,
if the haplotypes are loaded into a , we can easily import them straight
into the cache using .</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/CacheHaplotypes.html">CacheHaplotypes</a></span><span class="op">(</span><span class="va">SmallHaps</span><span class="op">)</span></span></code></pre></div>
<p>From we can inspect the haplotype cache to ensure that the data has
loaded correctly using . With the haplotypes loaded, next we must
specify our model parameters. First we supply a recombination map,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>m</mi><annotation encoding="application/x-tex">m</annotation></semantics></math>,
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>m</mi><mo>ℓ</mo></msup><annotation encoding="application/x-tex">m^\ell</annotation></semantics></math>
is the distance between variant
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mo>ℓ</mo><annotation encoding="application/x-tex">\ell</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>ℓ</mo><mo>+</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">\ell + 1</annotation></semantics></math>
in Morgans. Here we randomly generate a recombination map and set the
remaining parameters to their default values.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Beta.html" class="external-link">rbeta</a></span><span class="op">(</span><span class="fl">400</span><span class="op">-</span><span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">10</span><span class="op">)</span><span class="op">*</span><span class="fl">1e-6</span></span>
<span><span class="va">pars</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Parameters.html">Parameters</a></span><span class="op">(</span><span class="fu"><a href="../reference/CalcRho.html">CalcRho</a></span><span class="op">(</span>cM <span class="op">=</span> <span class="va">m</span>, s <span class="op">=</span> <span class="fl">1</span>, gamma <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, mu <span class="op">=</span> <span class="fl">1e-8</span><span class="op">)</span></span></code></pre></div>
<p>Next we must initialize a forward table object with our parameters
that will start at variant
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>0</mn><annotation encoding="application/x-tex">0</annotation></semantics></math>
and a backward table object that will start at the variant
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>L</mi><mo>+</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">L+1</annotation></semantics></math>.
We then use the function to propagate the forward table using parameters
to target variant
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>250</mn><annotation encoding="application/x-tex">250</annotation></semantics></math>.
We do the same with the backward table using the function. Note that and
do not return anything, the tables and are updated in place.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fwd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MakeForwardTable.html">MakeForwardTable</a></span><span class="op">(</span><span class="va">pars</span><span class="op">)</span></span>
<span><span class="va">bck</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MakeBackwardTable.html">MakeBackwardTable</a></span><span class="op">(</span><span class="va">pars</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/Forward.html">Forward</a></span><span class="op">(</span><span class="va">fwd</span>, <span class="va">pars</span>, <span class="fl">250</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Backward.html">Backward</a></span><span class="op">(</span><span class="va">bck</span>, <span class="va">pars</span>, <span class="fl">250</span><span class="op">)</span></span></code></pre></div>
<p>Now that and are at the same variant, they can be combined to obtain
the posterior marginal copying probabilities,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>p</mi><mo>ℓ</mo></msup><annotation encoding="application/x-tex">p^\ell</annotation></semantics></math>,
or a matrix of distances based on the negative log of those copying
probabilities,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>d</mi><mo>ℓ</mo></msup><annotation encoding="application/x-tex">d^\ell</annotation></semantics></math>,
as follows.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PostProbs.html">PostProbs</a></span><span class="op">(</span><span class="va">fwd</span>,<span class="va">bck</span><span class="op">)</span></span>
<span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DistMat.html">DistMat</a></span><span class="op">(</span><span class="va">fwd</span>,<span class="va">bck</span><span class="op">)</span></span></code></pre></div>
<p>The distance matrix can be easily clustered and visualized just by
calling</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">d</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in levelplot.formula(form, data, aspect = aspect, xlim = xlim, ylim =</span></span>
<span><span class="co">#&gt; ylim, : device support for raster images unknown, ignoring 'raster=TRUE'</span></span></code></pre></div>
<p><img src="Decoding_a_Single_Variant_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>In some applications, one might want to symmetrize the matrix before
plotting as follows.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fl">0.5</span><span class="op">*</span><span class="op">(</span><span class="va">d</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">d</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">d</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in levelplot.formula(form, data, aspect = aspect, xlim = xlim, ylim =</span></span>
<span><span class="co">#&gt; ylim, : device support for raster images unknown, ignoring 'raster=TRUE'</span></span></code></pre></div>
<p><img src="Decoding_a_Single_Variant_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Louis Aslett, Ryan Christ.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
