<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Reproducing kalis Paper LCT Example • kalis</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/PT_Serif-0.4.9/font.css" rel="stylesheet">
<link href="../deps/JetBrains_Mono-0.4.9/font.css" rel="stylesheet">
<link href="../deps/Roboto_Slab-0.4.9/font.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Reproducing kalis Paper LCT Example">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">kalis</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Decoding_a_Single_Variant.html">Decoding a Single Variant</a></li>
    <li><a class="dropdown-item" href="../articles/lct_example.html">Reproducing kalis Paper LCT Example</a></li>
  </ul>
</li>
<li class="nav-item"><a class="external-link nav-link" href="https://doi.org/10.1186/s12859-024-05688-8">BMC Bioinformatics Paper</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/louisaslett/kalis/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Reproducing kalis Paper LCT Example</h1>
                        <h4 data-toc-skip class="author">Ryan R. Christ
and Louis J. M. Aslett</h4>
            
            <h4 data-toc-skip class="date">11 December 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/louisaslett/kalis/blob/master/vignettes/lct_example.Rmd" class="external-link"><code>vignettes/lct_example.Rmd</code></a></small>
      <div class="d-none name"><code>lct_example.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview-accompanying-data">Overview &amp; Accompanying Data<a class="anchor" aria-label="anchor" href="#overview-accompanying-data"></a>
</h2>
<p>This vignette reproduces the distance matrix included in the initial
paper introducing our R package kalis. We infer the haplotype structure
at the lactase gene (LCT) in 1000 Genomes Phase 3 data.</p>
<p>Four data files are required. One is the recombination map
<code>lct.map</code> that accompanies this vignette, although another
recombination map of the user’s choosing may be used. Three files
encoding the required 1000 Genomes haplotypes, <code>lct.hap.gz</code> ,
<code>lct.legend.gz</code>, and <code>lct.sample</code>. All four files
can be found in the <code>example</code> folder at Github repository <a href="https://github.com/louisaslett/kalis-bmc" class="external-link">louisaslett/kalis-bmc</a>.
If the user would like to reconstruct these files directly from 1000
Genomes data, we we provide instructions for doing so using
<code>bcftools</code> at the end of this vignette.</p>
</div>
<div class="section level2">
<h2 id="run-kalis-in-r">Run kalis in R<a class="anchor" aria-label="anchor" href="#run-kalis-in-r"></a>
</h2>
<p>Start by declaring the number of cores available for kalis to use in
this example (<code>nthreads</code>) and the data directory that should
be used, <code>data_dir</code>. This data directory must contain the
recombination map <code>lct.map</code>, <code>lct.hap.gz</code> ,
<code>lct.legend.gz</code>, and <code>lct.sample</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nthreads</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">data_dir</span> <span class="op">&lt;-</span> <span class="st">"./"</span></span></code></pre></div>
<p>No further changes should be needed to the following for
replication.</p>
<p>We load our required libraries for this vignette.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://kalis.louisaslett.com/">"kalis"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://henrikbengtsson.github.io/R.utils/" class="external-link">"R.utils"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://r-datatable.com" class="external-link">"data.table"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://danifold.net/fastcluster.html" class="external-link">"fastcluster"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://Matrix.R-forge.R-project.org" class="external-link">"Matrix"</a></span><span class="op">)</span> </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://sjmgarnier.github.io/viridisLite/" class="external-link">"viridisLite"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/stephenturner/kgp" class="external-link">"kgp"</a></span><span class="op">)</span></span></code></pre></div>
<p>Now we’re ready to run kalis.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Declare LS Model Parameters</span></span>
<span><span class="co">#########################################</span></span>
<span><span class="va">neg_log10_Ne</span> <span class="op">&lt;-</span> <span class="fl">10</span></span>
<span><span class="va">neg_log10_mu</span> <span class="op">&lt;-</span> <span class="fl">4</span></span>
<span></span>
<span><span class="co"># Declare Target Locus</span></span>
<span><span class="co">#########################################</span></span>
<span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="st">"lct"</span></span>
<span><span class="va">gene_target_pos</span> <span class="op">&lt;-</span> <span class="fl">136608646</span> <span class="co"># rs4988235 in hg19 coordinates</span></span>
<span><span class="va">pos</span> <span class="op">&lt;-</span> <span class="fu">fread</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="va">gene</span>, <span class="st">".legend.gz"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">position</span></span>
<span><span class="va">target_idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="va">pos</span> <span class="op">&gt;=</span> <span class="va">gene_target_pos</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run kalis</span></span>
<span><span class="co">#########################################</span></span>
<span></span>
<span><span class="fu"><a href="../reference/CacheHaplotypes.html">CacheHaplotypes</a></span><span class="op">(</span>haps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="va">gene</span>, <span class="st">".hap.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">diff_map</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="fu">fread</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="va">gene</span>, <span class="st">".map"</span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">pars</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Parameters.html">Parameters</a></span><span class="op">(</span>rho <span class="op">=</span> <span class="fu"><a href="../reference/CalcRho.html">CalcRho</a></span><span class="op">(</span><span class="va">diff_map</span>, s <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="va">neg_log10_Ne</span><span class="op">)</span>, mu <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="va">neg_log10_mu</span><span class="op">)</span></span>
<span><span class="va">fwd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MakeForwardTable.html">MakeForwardTable</a></span><span class="op">(</span><span class="va">pars</span><span class="op">)</span></span>
<span><span class="va">bck</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MakeBackwardTable.html">MakeBackwardTable</a></span><span class="op">(</span><span class="va">pars</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/Forward.html">Forward</a></span><span class="op">(</span><span class="va">fwd</span>, <span class="va">pars</span>, <span class="va">target_idx</span>, nthreads <span class="op">=</span> <span class="va">nthreads</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Backward.html">Backward</a></span><span class="op">(</span><span class="va">bck</span>, <span class="va">pars</span>, <span class="va">target_idx</span>, nthreads <span class="op">=</span> <span class="va">nthreads</span><span class="op">)</span></span>
<span></span>
<span><span class="va">M</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DistMat.html">DistMat</a></span><span class="op">(</span><span class="va">fwd</span>, <span class="va">bck</span>, type <span class="op">=</span> <span class="st">"raw"</span>, nthreads <span class="op">=</span> <span class="va">nthreads</span><span class="op">)</span></span></code></pre></div>
<p>Note <code>M</code> may be plotted directly by calling
<code>plot(M)</code>.</p>
</div>
<div class="section level2">
<h2 id="post-processing-remake-paper-figure">Post-processing remake paper figure<a class="anchor" aria-label="anchor" href="#post-processing-remake-paper-figure"></a>
</h2>
<p>Here we provide post-processing code to reproduce the heatmap in our
paper from the distance matrix <code>M</code>. We start with a few
helper functions.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot_mat</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">file</span>, <span class="va">raster</span> <span class="op">=</span> <span class="cn">TRUE</span>, <span class="va">rel_scale</span> <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>  <span class="va">temp_col_scale</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu">viridisLite</span><span class="fu">::</span><span class="fu">viridis</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="va">rel_scale</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">mx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">ceiling</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">x</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="kw">if</span><span class="op">(</span><span class="va">mx</span> <span class="op">&gt;</span> <span class="fl">100</span><span class="op">)</span> <span class="op">{</span> <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="st">"the max entry of x cannot exceed 100 for this plot's color scale"</span><span class="op">)</span> <span class="op">}</span></span>
<span>    <span class="va">temp_col_scale</span> <span class="op">&lt;-</span> <span class="va">temp_col_scale</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">mx</span><span class="op">]</span></span>
<span>  <span class="op">}</span></span>
<span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/grDevices/cairo.html" class="external-link">cairo_pdf</a></span><span class="op">(</span><span class="va">file</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">lattice</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/lattice/man/levelplot.html" class="external-link">levelplot</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">:</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>                           useRaster <span class="op">=</span> <span class="va">raster</span>,</span>
<span>                           col.regions <span class="op">=</span> <span class="fu">grDevices</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette</a></span><span class="op">(</span><span class="va">temp_col_scale</span><span class="op">)</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>                           yaxt <span class="op">=</span> <span class="st">"n"</span>, xaxt <span class="op">=</span> <span class="st">"n"</span>, xlab <span class="op">=</span> <span class="st">""</span>, ylab <span class="op">=</span> <span class="st">""</span>, xaxt <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">interp_hapmap</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">path</span>,<span class="va">bp</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">d</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu">fread</span><span class="op">(</span><span class="va">path</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/approxfun.html" class="external-link">approx</a></span><span class="op">(</span><span class="va">d</span><span class="op">$</span><span class="va">`Position(bp)`</span>, <span class="va">d</span><span class="op">$</span><span class="va">`Map(cM)`</span>, xout <span class="op">=</span> <span class="va">bp</span>, method <span class="op">=</span> <span class="st">"linear"</span>, rule <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">$</span><span class="va">y</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">dip2hapidx</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">x</span> <span class="op">&lt;-</span> <span class="fl">2</span><span class="op">*</span><span class="va">x</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">x</span><span class="op">-</span><span class="fl">1</span>, <span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Now we can cluster <code>M</code> separately for African and
non-African haplotypes.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load sample population information</span></span>
<span><span class="co">#########################################</span></span>
<span><span class="va">cluster_by</span> <span class="op">&lt;-</span> <span class="st">"isAFR"</span></span>
<span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu">fread</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="va">gene</span>, <span class="st">".sample"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span></span>
<span><span class="va">init_order_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="fu">data.table</span><span class="op">(</span><span class="st">"id"</span> <span class="op">=</span> <span class="va">id</span><span class="op">)</span>, <span class="va">kgp3</span>, by <span class="op">=</span> <span class="st">"id"</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">samples</span><span class="op">)</span> <span class="op">!=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span> <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="st">"some samples have been removed by merging with kgp3"</span><span class="op">)</span> <span class="op">}</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/all.equal-methods.html" class="external-link">all.equal</a></span><span class="op">(</span><span class="va">init_order_samples</span>, <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">samples</span><span class="op">$</span><span class="va">id</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span> <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="st">"some samples have been moved out of the order in lct.sample"</span><span class="op">)</span> <span class="op">}</span></span>
<span><span class="va">samples</span><span class="op">[</span>,<span class="va">isAFR</span> <span class="op">:=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">reg</span> <span class="op">==</span> <span class="st">"AFR"</span>, <span class="st">"AFR"</span>, <span class="st">"not_AFR"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Symmeterize &amp; Scale Distance Matrix at LCT</span></span>
<span><span class="co">###############################################</span></span>
<span><span class="va">M</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fl">0.5</span><span class="op">/</span><span class="op">(</span><span class="va">neg_log10_mu</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">*</span> <span class="op">(</span><span class="va">M</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Permute &amp; Cluster Distance Matrix</span></span>
<span><span class="co">###################################################################</span></span>
<span><span class="va">diploid_perm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">samples</span><span class="op">$</span><span class="va">reg</span>, <span class="va">samples</span><span class="op">$</span><span class="va">pop</span>, <span class="va">samples</span><span class="op">$</span><span class="va">id</span><span class="op">)</span></span>
<span><span class="va">psamples</span> <span class="op">&lt;-</span> <span class="va">samples</span><span class="op">[</span><span class="va">diploid_perm</span>,<span class="op">]</span></span>
<span></span>
<span><span class="va">haploid_perm</span> <span class="op">&lt;-</span> <span class="fu">dip2hapidx</span><span class="op">(</span><span class="va">diploid_perm</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pM</span> <span class="op">&lt;-</span> <span class="va">M</span><span class="op">[</span>, <span class="va">haploid_perm</span><span class="op">]</span><span class="op">[</span><span class="va">haploid_perm</span>,<span class="op">]</span></span>
<span></span>
<span><span class="va">hap_groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">psamples</span><span class="op">[[</span><span class="va">cluster_by</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">hap_groups</span> <span class="op">&lt;-</span> <span class="va">hap_groups</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">psamples</span><span class="op">[[</span><span class="va">cluster_by</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">baseline_idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html" class="external-link">cumsum</a></span><span class="op">(</span><span class="fl">2</span><span class="op">*</span><span class="va">hap_groups</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">hap_groups</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">baseline_idx</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">hap_groups</span><span class="op">)</span></span>
<span></span>
<span><span class="va">order_M</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">as.list</a></span><span class="op">(</span><span class="va">hap_groups</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">order_M</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">hap_groups</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">hap_groups</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">current_pop_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">psamples</span><span class="op">[[</span><span class="va">cluster_by</span><span class="op">]</span><span class="op">]</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">hap_groups</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">current_pop_haplotypes</span> <span class="op">&lt;-</span> <span class="fu">dip2hapidx</span><span class="op">(</span><span class="va">current_pop_samples</span><span class="op">)</span></span>
<span>  <span class="va">sM</span> <span class="op">&lt;-</span> <span class="va">pM</span><span class="op">[</span><span class="va">current_pop_haplotypes</span>, <span class="va">current_pop_haplotypes</span><span class="op">]</span></span>
<span>  <span class="va">order_M</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">hap_groups</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">baseline_idx</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">hap_groups</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="fu">fastcluster</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/fastcluster/man/hclust.html" class="external-link">hclust</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">as.dist</a></span><span class="op">(</span><span class="va">sM</span><span class="op">)</span>, method<span class="op">=</span><span class="st">"average"</span><span class="op">)</span><span class="op">$</span><span class="va">order</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">order_M</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">order_M</span><span class="op">)</span></span>
<span><span class="va">cM</span> <span class="op">&lt;-</span> <span class="va">pM</span><span class="op">[</span>, <span class="va">order_M</span><span class="op">]</span><span class="op">[</span><span class="va">order_M</span>,<span class="op">]</span></span>
<span></span>
<span></span>
<span><span class="co"># Plot clustered Distance Matrix </span></span>
<span><span class="co">#########################################</span></span>
<span><span class="fu">plot_mat</span><span class="op">(</span><span class="va">cM</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="va">gene</span>, <span class="st">"_dist_mat.pdf"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="reconstructing-lct-files-directly-from-1000-genomes-data-with-bcftools">Reconstructing LCT files directly from 1000 Genomes data with
bcftools<a class="anchor" aria-label="anchor" href="#reconstructing-lct-files-directly-from-1000-genomes-data-with-bcftools"></a>
</h2>
<p>Download a phased VCF of Chromosome 2 from the 1000 Genomes website,
<code>https://www.internationalgenome.org/</code>. Below, we assume that
VCF has name <code>chr2.vcf.gz</code>. The 1000 Genomes Project
periodically makes updates to the VCFs available. We used
<code>ALLchr2phase3_shapeit2_mvncall_integrated_v5a20130502genotypes.vcf.gz</code>
in our analysis. That data was in hg19 coordinates, so throughout this
vignette, we work entirely in hg19 coordinates.</p>
<p>Please be mindful that some modification of the recombination map and
target locus index will be necessary if working with data based on a
different build.</p>
<p>If <code>bcftools</code> is not already installed, first install
<code>bcftools</code> from
<code>https://samtools.github.io/bcftools/</code>. Then from a bash
terminal, run</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">--regions</span> 2:136608646-136608646 <span class="at">--types</span> snps <span class="at">--min-ac</span> 2:minor <span class="at">-Ou</span> <span class="at">--threads</span> 1 chr2.vcf.gz <span class="kw">|</span> <span class="ex">bcftools</span> convert <span class="at">-h</span> lct <span class="at">--threads</span> 1</span></code></pre></div>
<p>Note, both of the above commands take a <code>--threads</code>
argument. Here, we’ve set it to <code>1</code> but this can increased to
the number of available cores to increase the execution speed. See the
<code>bcftools</code> documentation for more details.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Louis Aslett, Ryan Christ.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
